src/rootfs.rs

```diff
@@ -387,3 +392,55 @@ fn parse_mount(m: &Mount) -> (MsFlags, String) {
     }
     (flags, data.join(","))
 }
+
+fn get_parent_mount(rootfs: &Path) -> Result<MountInfo> {
+    let mount_infos = Process::myself()?.mountinfo()?;
+    // find the longest mount point
+    let parent_mount_info = mount_infos
+        .into_iter()
+        .filter(|mi| rootfs.starts_with(&mi.mount_point))
+        .max_by(|mi1, mi2| mi1.mount_point.len().cmp(&mi2.mount_point.len()))
+        .ok_or_else(|| anyhow!("couldn't find parent mount of {}", rootfs.display()))?;
+    Ok(parent_mount_info)
+}
+
+/// Make parent mount of rootfs private if it was shared, which is required by pivot_root.
+/// It also makes sure following bind mount does not propagate in other namespaces.
+fn make_parent_mount_private(rootfs: &Path) -> Result<()> {
+    let parent_mount = get_parent_mount(rootfs)?;
+
+    // check parent mount has 'shared' propagation type
+    if parent_mount
+        .opt_fields
+        .iter()
+        .any(|field| matches!(field, MountOptFields::Shared(_)))
+    {
+        nix_mount(
+            None::<&str>,
+            &parent_mount.mount_point,
+            None::<&str>,
+            MsFlags::MS_PRIVATE,
+            None::<&str>,
+        )?;
+    }
+
+    Ok(())
+}
+
+/// Change propagation type of rootfs as specified in spec.
+pub fn adjust_root_mount_propagation(spec: &Spec) -> Result<()> {
```

Comment:
Can we pass in linux rather than the whole spec. Looks like only the linux is used.